pkg/strategy/dca2/state.go

```diff
@@ -0,0 +1,202 @@
+package dca2
+
+import (
+	"context"
+	"time"
+)
+
+type State int64
+
+const (
+	None State = iota
+	WaitToOpenPosition
+	PositionOpening
+	OpenPositionReady
+	OpenPositionOrderFilled
+	OpenPositionOrdersCancelling
+	OpenPositionOrdersCancelled
+	TakeProfitReady
+)
+
+func (s *Strategy) initializeNextStateC() bool {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	isInitialize := false
+	if s.nextStateC == nil {
+		s.logger.Info("[DCA] initializing next state channel")
+		s.nextStateC = make(chan State, 1)
+	} else {
+		s.logger.Info("[DCA] nextStateC is already initialized")
+		isInitialize = true
+	}
+
+	return isInitialize
+}
+
+func (s *Strategy) emitNextState(nextState State) {
+	select {
+	case s.nextStateC <- nextState:
+	default:
+		s.logger.Info("[DCA] nextStateC is full or not initialized")
+	}
+}
+
+// runState
+// WaitToOpenPosition -> after startTimeOfNextRound, place dca orders ->
+// PositionOpening
+// OpenPositionReady -> any dca maker order filled ->
+// OpenPositionOrderFilled -> price hit the take profit ration, start cancelling ->
+// OpenPositionOrdersCancelled -> place the takeProfit order ->
+// TakeProfitReady -> the takeProfit order filled ->
+func (s *Strategy) runState(ctx context.Context) {
+	s.logger.Info("[DCA] runState")
+	ticker := time.NewTicker(5 * time.Second)
```

Comment:
why do we use 5 seconds here?

Reply:
This ticker is used to check if the state works well.
Because some state like WaitToOpen, we need to use it to know when to open position